pkg/receive/handler.go

```diff
@@ -479,5 +504,47 @@ func isConflict(err error) bool {
 	if err == nil {
 		return false
 	}
-	return err == conflictErr || err == storage.ErrDuplicateSampleForTimestamp || err == storage.ErrOutOfOrderSample || err == storage.ErrOutOfBounds || err.Error() == strconv.Itoa(http.StatusConflict)
+	return err == conflictErr ||
+		err == storage.ErrDuplicateSampleForTimestamp ||
+		err == storage.ErrOutOfOrderSample ||
+		err == storage.ErrOutOfBounds ||
+		err.Error() == strconv.Itoa(http.StatusConflict) ||
+		status.Code(err) == codes.AlreadyExists
+}
+
+func newPeerGroup(dialOpts ...grpc.DialOption) *peerGroup {
+	return &peerGroup{
+		dialOpts: dialOpts,
+		cache:    map[string]storepb.WriteableStoreClient{},
+		m:        sync.Mutex{},
+		dialer:   grpc.DialContext,
+	}
+}
+
+type peerGroup struct {
+	dialOpts []grpc.DialOption
+	cache    map[string]storepb.WriteableStoreClient
+	m        sync.Mutex
+
+	// dialer is used for testing.
+	dialer func(ctx context.Context, target string, opts ...grpc.DialOption) (conn *grpc.ClientConn, err error)
+}
+
+func (p *peerGroup) Get(ctx context.Context, addr string) (storepb.WriteableStoreClient, error) {
```

Comment:
can we keep this unexported?

Reply:
Yeah we can change it. The struct isn't exported so it doesn't change much though.